Support module exports

ahoym opened the pull request:
Hello, thanks for writing this jQuery-less smooth scroll - it's awesome!

I wanted to use this for a small project, but I noticed this library wasn't module export-able. I bundle my `js` files with browserify, which allows you to `require` modules like in node.

This PR:

+ Creates simple checks for whether the environment is a browser or node. If the former, smoothScroll will know it is being used in the `<script>` context and carry on as usual. If the latter, smoothScroll will become exportable and can be required as a module.
+ Converts direct references to the `window` to now use the `globalWindow` variable. In both browser and node, `globalWindow` should still refer to the `window`.
+ Adds a new minified file of the new script.

The `<script>` in the `example.html` still works, and if you want to check this out in action as a require-able module, you can do:
```
npm install --save "git://github.com/ahoym/smoothScroll.git#support_modules_export"
```

and it should work in your browserified files. I confirmed this with my own project.

Comments and suggestions are of course welcome

t1mmen commented:
@ahoym Will this work with server-side too? If so, this looks like a much better approach than what I'm doing in https://github.com/alicelieutier/smoothScroll/pull/11

alicelieutier commented on smoothscroll.js:
```diff
@@ -78,8 +83,16 @@ document.addEventListener("DOMContentLoaded", function () {
 
-})();
-
+})(this);
```
I'm probably missing something here, but what are you calling the function with `this`?

ahoym replied:
Nope you're right, I accidentally missed a step here. This should either be:
```
(function() {
var root = this;
...
}).call(this);
```

or
```
(function(root) {
...
})(this);
```

Which do you prefer? Either way I left this comment at the top:
```javascript
// In <script> tag context, `this` is the `window`.
// In Node.js context, `this` is the node `global`.
```

https://github.com/ahoym/smoothScroll/blob/support_modules_export/smoothscroll.js#L3-L4

alicelieutier replied:
I think the first one makes it a little bit clearer.

ahoym commented:
@t1mmen, 

I can't say 100% for sure, but I'm confident it would. I'm using a bundler, browserify, which allows users to use the node-style `require`, and my branch works fine in my mini poject. 

If you could confirm it with your server-side code, that would be really helpful. You can install my branch of `smoothScroll` with:

```
npm install --save "git://github.com/ahoym/smoothScroll.git#support_modules_export"
```

Thanks!

t1mmen commented:
My usage, in a React component, is:

`import smoothScroll from 'smoothscroll'

It fails server-side with your code, with the following errors:

```
/srv/app/web/node_modules/smoothscroll/smoothscroll.js:9
if(document.querySelectorAll === void 0 || globalWindow.pageYOffset === void 0
   ^
ReferenceError: document is not defined
    at /srv/app/web/node_modules/smoothscroll/smoothscroll.js:9:4
    at Object.<anonymous> (/srv/app/web/node_modules/smoothscroll/smoothscroll.js:98:3)
    at Module._compile (module.js:446:26)
    at Module._extensions..js (module.js:464:10)
    at Object.require.extensions.(anonymous function) [as .js] (/srv/app/web/node_modules/babel/lib/babel/api/register/node.js:130:7)
    at Module.load (module.js:341:32)
    at Function.Module._load (module.js:296:12)
    at Module.require (module.js:351:17)
    at require (module.js:370:17)
    at Object.<anonymous> (/srv/app/web/app/mixins/smoothScrolling.js:3:20)
```

You should get past this by doing something like this: https://github.com/alicelieutier/smoothScroll/pull/11/files#diff-9ab594b6f980ebeae044625095865ae0R4

ahoym commented:
> ReferenceError: document is not defined

Cool, I thought this might blow up after I looked at your PR. 

In this case `document` should be `globalWindow.document` because on the server side, `document` isn't defined. However we can get around this because `document` is actually on the `window` object. I'll make a quick fix for this!

t1mmen commented:
@ahoym, still errors out, I'm afraid. I'm pretty sure this is because server-side, node doesn't know about `window` or `document`. 

Compile error:
```
/srv/app/web/node_modules/smoothscroll/smoothscroll.js:9
if(globalWindow.document.querySelectorAll === void 0 || globalWindow.pageYOffs
               ^
TypeError: Cannot read property 'document' of undefined
    at /srv/app/web/node_modules/smoothscroll/smoothscroll.js:9:16
    at Object.<anonymous> (/srv/app/web/node_modules/smoothscroll/smoothscroll.js:98:3)
    at Module._compile (module.js:446:26)
    at Module._extensions..js (module.js:464:10)
    at Object.require.extensions.(anonymous function) [as .js] (/srv/app/web/node_modules/babel/lib/babel/api/register/node.js:130:7)
    at Module.load (module.js:341:32)
    at Function.Module._load (module.js:296:12)
    at Module.require (module.js:351:17)
    at require (module.js:370:17)
    at Object.<anonymous> (/srv/app/web/app/mixins/smoothScrolling.js:3:20)
```

ahoym commented:
Hm, that's unfortunate. Looking at your PR again, how are you using `smoothScroll`? From what it looks like, it doesn't instantiate when you're on the server? Does this mean you also have it in the client somewhere?

Sorry, I haven't written an isomorphic app yet, so any knowledge you can pass down would be extremely helpful!

If anything, i'm thinking doing something like this (like you suggested) might solve our problem:
```javascript
(function(root){
// In <script> tag context, `root` is the `window`.
// In Node.js context, `root` is the node `global`.
var globalWindow = typeof global === 'undefined' ? root : root.window;

// Do no initialize when running server side:
if (typeof globalWindow !== 'object' || typeof globalWindow.document !== 'object') return;

...
```

t1mmen commented:
I got it working by using your latest code, but modifying it to start off with:

```
(function(){

// Do not initialize when running server side:
if (typeof window !== 'object' || typeof document !== 'object') return;

var root = this; // etc, etc..
```

Be ware, that the usage of the plugin is changed with this PR: `window.smoothScroll(target, duration, callback)` no longer works (for me, anyway), instead I needed to do just `smoothScroll(target, duration, callback)`

> From what it looks like, it doesn't instantiate when you're on the server? Does this mean you also have it in the client somewhere?

Correct, it doesn't instantiate on the server, only on the client. The goal of the `if (typeof window !== 'object' || typeof document !== 'object') return;` part, is to abort the instantiate on the server, but still execute on the client.

I'm using this in a React project with, via a Mixin. 

#### Mixin: 
```
'use strict';

import smoothScroll from 'smoothscroll';

const smoothScrolling = {

  toId: function(selector) {
    if (typeof window !== 'object') return;
    const target = document.getElementById(selector);
    smoothScroll(target); // NB! Used to be window.smoothScoll(target) before this PR
  }

}

export default {
  smoothScrolling: smoothScrolling // to scope the mixin
}
```

#### Usage of mixin in React components
```
'use strict';
import React from 'react';
import SmoothScrolling from 'mixins/smoothScrolling';

export default React.createClass({
  mixins: [ SmoothScrolling ],

  doSomething(event) {
    event.preventDefault();
    this.smoothScrolling.toId('content');
  },

  render() {
    return (<button onClick={this.doSomething}>Do the scroll thing</button>);
  }
});
```

ahoym commented:
Cool, I think I fully understand your situation now. I'll cherry-pick your commit and update mine's accordingly.

> Be ware, that the usage of the plugin is changed with this PR: window.smoothScroll(target, duration, callback) no longer works (for me, anyway), instead I needed to do just smoothScroll(target, duration, callback)

Yes, if you're importing it as a module, it shouldn't pollute the `window`.

ahoym commented:
This PR is getting really crazy and inflated with unnecessary commits. I propose making this a test branch until we can find a working solution that addresses all our needs. Then once we find one, I can open up a new PR that only contains clean commits, so that the git history wont be super cluttered.

@alicelieutier, @t1mmen, here's what I have so far now:

+ Changed my crude check of the environment into a full blown UMD wrapper. This should work for the global `<script>` context, RequireJS, and CommonJS module formats. Here's a great article about the wrapper itself, for those that are interested: http://toddmotto.com/what-function-window-document-undefined-iife-really-means/
+ Added @t1mmen's block of instantiation on the server. Note, I added it to the wrapper and not in the library. I think it should work but if you could confirm that would be awesome!
+ Reverted all `globalWindow` jargon back to the original state of the library.
+ Added a `'use strict';` to prevent further confusion. 

`this` in non-strict functions refers to the global object, which was enabling fragile behavior (the `var root = this;` worked mainly because of this reason, unknowingly to me).

t1mmen commented:
I can confirm that the current version of this PR supports isomorphic/universal usage :+1:

alicelieutier commented:
I like this solution, it's super neat to have one place at the beginning to manage all the different cases based on the context.
However, in this case, the line:
```
if (typeof window !== 'object' || typeof document !== 'object') return;
```
basically means that everything will only execute in the context of a browser, or at least a context that has a window and a document. So instead of passing root, it would be clearer, in this specific case, to pass `window` and `document`, and then attach the function to window.

ahoym commented:
So something along the lines of:

```javascript
(function (window, smoothScroll) {
  'use strict';

  // Support RequireJS and CommonJS/NodeJS module formats.
  // Attach smoothScroll to the `window` when executed as a <script>.

  // RequireJS
  if (typeof define === 'function' && define.amd) {
    define(smoothScroll);

  // CommonJS
  } else if (typeof exports === 'object' && typeof module === 'object') {
    module.exports = smoothScroll();
 
  // <script>
  } else if (typeof window === 'object') {
    window.smoothScroll = smoothScroll();

  // Do not initialize smoothScroll when running server side, handle it in client:
  } else {
    return;
  }

})(window, function(){
  'use strict';
  ...
```

I took the liberty of not passing in the `document` to the wrapper since we don't use it. The wrapper will `return` if none of the conditions are met, which handles the case that @t1mmen needs.

alicelieutier commented:
LGTM. Does this work as intended for you?

@t1mmen was there a specific reason you were testing for both `window` and `document`?

If this is browser context, anyway, `document` would be a member of `window` anyway, so you could only pass `window` in and then test for both `window` and `window.document`.

t1mmen commented:
@alicelieutier No, that was mostly me being overly cautious. Just checking `window` is fine, I'm sure.

alicelieutier commented:
Then I guess we are all happy we this version :)
@ahoym would you mind making a unified PR that references to this conversation?

Thanks so much for working on this!

ahoym commented:
@alicelieutier, made!

Closing in favor of: https://github.com/alicelieutier/smoothScroll/pull/13
